### ISSUE TITLE
`@SerialComment` annotation

### ISSUE THREAD

**Issue by slava110** [Kotlin/kotlinx.serialization#1970]
## What is your use-case and why do you need this feature?
`@SerialComment` annotation will be useful for automatically generated configuration files. Some serialization libraries already implemented that feature including:
- [akuleshov7/ktoml](https://github.com/akuleshov7/ktoml/blob/main/ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/annotations/TomlComments.kt)
- [charleskorn/kaml](https://github.com/charleskorn/kaml/blob/main/src/commonMain/kotlin/com/charleskorn/kaml/YamlComment.kt)
- [Him188/yamlkt](https://github.com/Him188/yamlkt/blob/master/yamlkt/src/commonMain/kotlin/net.mamoe.yamlkt/YamlComment.kt)

And it's possible to add comment support in other libraries like:
- [kotlinx.serialization.properties](https://github.com/Kotlin/kotlinx.serialization/tree/master/formats/properties)
- [kotlinx.serialization.hocon](https://github.com/Kotlin/kotlinx.serialization/tree/master/formats/hocon)
- [adamko-dev/json5-kotlin](https://github.com/adamko-dev/json5-kotlin)

The problem is that they are using/will use different `@Comment` annotations
## Describe the solution you'd like
- Add `@SerialComment` annotation
- `SerialDescriptor.getElementComment(index: Int): String` function to get property comment
~~`SerialFormat.supportsComments` property to check if `SerialFormat` supports comment serialization (`false` by default)~~

**Comment by slava110** [comment]
Yep, maybe there's no point in `supportsComments`, let's forget about it

> It appears that you want to have the annotation be specified on types, not use sites/fields

Actually I wanted to have the annotation be specified on properties, not types, yeah (Updated my original comment)
I think comments should be written on each serialization (is it even possible to restrict it to first time only from `kotlinx.serialization`?..)

> Besides these questions I'm not sure what the exact use case is

Commented properties are useful for any type of automatically generated configuration files. Developers will be able to tell users what should be put in certain JSON properties for example

**Comment by pdvrieze** [comment]
@slava110 So the use case is really configuration files. I can see that one. It doesn't really suffer from the repetition issue much and I certainly recognize the value of explanations/instructions in configuration files.
As to first time only, the format would need to maintain a set with type/field information that has already been written.

**Comment by whyoleg** [comment]
BTW, even if protobuf doesn't support comments in messages, it could be useful for schema generation to have comments there.
Another thing, that should be considered, is that when we will have something like `SerialComment`, then it will be a little confusing, when we will have both description for property and similar/different text in `SerialComment`. Not an issue, but something we should keep in mind.
Ex.
```kotlin
/**
 * @property used for ...
 */
data class SomeData(
  @SerialComment("property used for ...")
  val property: String
)
```

**Comment by slava110** [comment]
@pdvrieze I think maintaining set of type/field information that has already written will add unnecessary complexity and will create more questions. It's an idea for completely different issue
Currently `kotlinx.serialization` formats are writing everything each time object serialized (at least most of them, maybe some aren't) so comments should be written on each serialization too

**Comment by vnermolaev** [comment]
I guess you can already achieve the goal as follows

```kotlin
@SerialInfo
@Target(AnnotationTarget.PROPERTY)
annotation class SerialComment(val comment: String)

fun SerialDescriptor.getElementComment(index: Int): String {
  val serialComment = getElementAnnotations(index).firstOrNull { SerialComment::class.isInstance(it) } as? SerialComment
  return serialComment?.comment ?: ""
}

@Serializable
data class Container(
    @SerialComment("Property is used for demonstration")
    val i: Int
)

fun main(args: Array<String>) {
    println(Container.serializer().descriptor.getElementComment(0))
}

```

**Comment by sgammon** [comment]
@whyoleg > BTW, even if protobuf doesn't support comments in messages

do you mean at runtime? protobuf does support comments in messages. actually this is my use case for finding this issue.

@pdvrieze perhaps my use case helps too. i want to preserve comments from Kotlin `data class` properties so that i can express them in generated `.proto` files. the proto files are used downstream and the comments are meaningful there -- they turn into code hints and other useful things that we want to keep in sync with the property comments on the source of truth (the kotlin data class).

**Comment by sgammon** [comment]
i wonder, though, if it is possible to do make this happen without a dedicated annotation. does the compiler plugin support comments? could it just take the native Kotlin comment from the property?

**Comment by solonovamax** [comment]
Is this issue being worked on at all?

I'd really love to see comment support in kotlinx.serialization

**Comment by sandwwraith** [comment]
@solonovamax No, it is not in our current plans.
